Source/DafnyCore/AST/Grammar/IFileSystem.cs

```diff
@@ -48,7 +48,10 @@ public class OnDiskFileSystem : IFileSystem {
   }
 
   public TextReader ReadFile(Uri uri) {
-    return new StreamReader(uri.LocalPath);
+    var reader = new StreamReader(uri.LocalPath);
+    var str = reader.ReadToEnd();
+    reader.Close();
+    return new StringReader(str);
```

Comment:
This is not OK for performance reasons. We need to design our code towards parsing in a streaming fashion.

Reply:
I usually worry more about correctness than performance, but I think I might find something that suits both of our needs and I'll put it up in a separate PR.

Reply:
I was about to write a StreamReader that auto-closes the file it reads when done, since I haven't found any. However, after analzying the code further, I don't think it is necessary.

Here is an analysis to consider:

* `IFileSystem.cs:50  public TextReader ReadFile(Uri uri) {` is called by
  * `DafnyProject.cs:53` which has a `ReadToEndAsync` but does not closes the stream
  * `IFileSystem.cs:25   public TextReader ReadFile(Uri uri) {` which has the same usage as its parents
  * `LanguageServerFileSystem.cs:41 existingText = OnDiskFileSystem.Instance.ReadFile(uri).ReadToEnd();` which reads to the end but does not closes the stream
  * `LanguageServerFileSystem:92` which has the same usage as its parents
  * `DafnyFile.cs:96` store the call as a closure in the `GetContent` of a field of a `DafnyFile`. This field is read in different places:
    * DafnyConsolePrinter.cs:72` `var reader = file.GetContent(); lines = Util.Lines(reader).ToList();`. Similarly, we read to the end to get all the lines but it does not close the stream
    * `ProgramParser.cs:269` `var text = SourcePreprocessor.ProcessDirectives(getReader(), new List<string>());` which returns a source file after preprocessing. as a single string, but does not close the stream
    * `CoverageReporter.cs:321` `var source = dafnyFile.GetContent().ReadToEnd();` reads to the end but does not close the stream
    * `FormatCommand.cs:75` `using var content = dafnyFile.GetContent(); var originalText = await content.ReadToEndAsync();` This one also reads to the end and close the stream appropriately (I don't remember doing that, probably you did when you migrate to the new file system, so thanks)

In short, we currently never rely on streaming files and just forget to close it in almost every place. C# does perform auto-closing but is lazy and for big files it won't do it fast enough, which crashes the formatter on the command-line randomly.
So my recommandation is just to do this read and close it immediately so that it becomes possible to write after read. I'll create a separate PR for that.

Reply:
https://github.com/dafny-lang/dafny/pull/4950

Reply:
> DafnyProject.cs:53 which has a ReadToEndAsync but does not closes the stream

What makes you say it does not close the stream? It uses a `using` directive:
```
using var textReader = fileSystem.ReadFile(uri);
```

Good finds though. Let's fix them!